<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Asynchronous JavaScript</title>
    <link rel="stylesheet" href="./resources/css/index.css">
    <link rel="stylesheet" href="../../../../index.css">
    <link rel="stylesheet" href="./resources/prismjs/prism.css">
</head>

<body>
    <header>
        <nav>
            <ul class="breadcrumb">
                <li><a href="../../../../">Home</a></li>
                <li><a href="../../">Reference Pages</a></li>
                <li><a href="../">JavaScript</a></li>
                <li><a href="#">Asynchronous JavaScript</a></li>
            </ul>
        </nav>
        <h1>Asynchronous JavaScript</h1>
    </header>
    <main>
        <section>
            <h2>Event Loop</h2>
            <p>JavaScript is <strong>single threaded</strong>, but uses the <strong>event loop</strong> to carry out
                async tasks.</p>
            <h3>Useful links</h3>
            <ul>
                <li><a href="https://www.javascripttutorial.net/javascript-event-loop/">javascripttutorial</a></li>
                <li><a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Event_loop">MDN</a></li>
            </ul>
            <p>When a program is run, function calls are made and added to a stack. The functions that make requests
                that need to wait for servers to respond then get sent to a separate queue. Once the stack has cleared,
                then the functions in the queue are executed. Very high level explanation of the event loop:</p>
            <ol>
                <li>The call stack executes functions in order</li>
                <li>When a function performs some async task that it needs to wait for, the Web API removes that
                    function
                    from the stack and adds it to the callback (or message) queue. For example, if
                    <strong>setTimeout</strong> is called, the web API sets a timer for when it will enqueue the
                    callback onto the <strong>message queue</strong> (they are stored on the heap).
                </li>
                <li>When the call stack is empty, the event loop will dequeue that function from the callback queue and
                    push it back onto the stack to execute.</li>
            </ol>
            <figure>
                <img src="./resources/images/javascript-event-loop-step-1.png"
                    alt="diagram that describes the JavaScript event loop">
                <figcaption><a href="https://www.javascripttutorial.net/javascript-event-loop/">source</a></figcaption>
            </figure>
        </section>
        <section>
            <h2>Promises</h2>
            <p>Promises are objects that represent the eventual outcome of an asynchronous operation. A Promise object
                can be in one of three states:</p>
            <ul>
                <li><strong>Pending</strong>: The initial stateâ€” the operation has not completed yet.</li>
                <li><strong>Fulfilled</strong>: The operation has completed successfully and the promise now has a
                    resolved value. For
                    example, a request's promise might resolve with a JSON object as its value.</li>
                <li><strong>Rejected</strong>: The operation has failed and the promise has a reason for the failure.
                    This reason is
                    usually an Error of some kind.</li>
            </ul>
            <article>
                <h3>Constructing a Promise</h3>
                <p>Below is how to create a new promise:</p>
                <pre><code class="language-javascript">const executorFunction = (resolve, reject) => {
    if (someCondition) {
        resolve('I resolved!');
    } else {
        reject('I rejected!'); 
    }
}
const myFirstPromise = new Promise(executorFunction);</code></pre>
                <p>A promise takes in a single <strong>executor function</strong> as input, which takes in two values:
                    <strong>resolve</strong> and <strong>reject</strong>.
                </p>
                <ul>
                    <li><strong>resolve</strong> is a function with
                        one argument. Under the hood, if invoked, resolve() will
                        change the promise's
                        status from pending to fulfilled, and the promise's resolved value will be set to the argument
                        passed
                        into resolve().</li>
                    <li><strong>reject</strong> is a function that takes a reason or error as an argument. Under the
                        hood,
                        if invoked, reject()
                        will change the promise's status from pending to rejected, and the promises rejection reason
                        will
                        be
                        set to the argument passed into reject().</li>
                </ul>
            </article>
            <article>
                <h3>Consuming Promises</h3>
                <p>When a promise is no longer pending, we can handle the outcome using the <strong>.then</strong>
                    method on
                    the promise.</p>
                <pre><code class="language-javascript">myFirstPromise.then(onFulfilled, onRejected);</code></pre>
                <p><strong>onFulfilled</strong> is the success handler function, which contains logic for handling a
                    successful promise. <strong>onRejected</strong> serves the opposite purpose.
                    <strong>onFulfilled</strong> takes as
                    argument what is passed into <strong>resolve</strong> in the promise, and vise-versa
                </p>
                <pre><code class="language-javascript">const onFulfilled = res => {
    // do something with res
}

const onRejected = rejectionReason => {
    ...
}
            </code></pre>
                <p><strong>.then()</strong> will return a promise with the same settled value as the promise it was
                    called
                    on if no appropriate handler was provided. This means that if the promise succeeds but no success
                    handler was provided, <strong>.then()</strong> will just return a promise with the same settled
                    results
                    as the original promise.
                    The same goes if no failure handler is provided.
                </p>
                <p>To write cleaner code and achieve <i>separation of concerns</i>, it is common to only include the
                    success
                    handler in the <strong>.then()</strong> method, and then chain a <strong>.catch()</strong> call:</p>
                <pre><code class="language-javascript">myFirstPromise.then((res) => {
    // onFulfilled function
    ...
}).catch((error) => {
    // onRejected function
};)
            </code></pre>
                <p><strong>.catch()</strong> is equivalent to a <strong>.then()</strong> call without the success
                    callback
                    provided:</p>
                <pre><code class="language-javascript">myFirstPromise.then((res) => {
    // onFulfilled function
    ...
}).then(null, (error) => {
    // onRejected function
};)</code></pre>
                <h3>Chaining Promises</h3>
                <p>It's not uncommon to have to perform a sequence of asynchronous tasks, each depending on the one
                    before
                    it. We can do this by chaining promises together, a process called <i>composition</i>. An example of
                    promise composition is shown below:</p>
                <pre><code class="language-javascript">myFirstPromise.then((firstPromiseResolvedValue) => {
    return mySecondPromise(secondPromisesolvedValue);
}).then((secondPromiseResolvedValue) => {
    return myThirdPromise(secondPromisesolvedValue);
}).then(((thirdPromiseResolvedValue) => {
    console.log(thirdPromiseResolvedValue);
}));
            </code></pre>
                <p>note that the <strong>return</strong> statement is necessary to pass the next promise through the
                    chain,
                    otherwise the default return value is used.</p>
            </article>
            <article>
                <h2>Promise.all()</h2>
                <p>The <strong>Promise.all()</strong> method is used to group a bunch of promises together and perform
                    some
                    operation when they are all completed. It takes in an array of promises, and if they all resolve,
                    Promise.all() will return a single promise that resolves with an array containing the resolve value
                    for
                    the initial promises. If any of the initial promises rejects, <strong>Promise.all()</strong> will
                    <i>fail fast</i> and return a promise with a rejection reason being why that one promise failed.
                </p>
                <pre><code class="language-javascript">const myPromises = Promise.all([ myFirstPromise, mySecondPromise, myThirdPromise ]);
myPromises.then(resolvedValues => {
    const myFirstPromiseResolvedValue = resolvedValues[0];
    const mySecondPromiseResolvedValue = resolvedValues[1];
    const myThirdPromiseResolvedValue = resolvedValues[2];
    ...
}).catch(rejectionReason => {
    console.log(rejectionReason);
});
            </code></pre>
            </article>
        </section>
        <section>
            <h2>Async / Await</h2>
            <p><strong>async... await</strong> is a new syntax that emerged in <strong>ES8</strong> as another way for
                handling asynchronous action</p>
            <article>
                <h3>The <u>async</u> Keyword</h3>
                The async keyword is used to write functions that contain asynchronous logic:
                <pre><code class="language-javascript">async function asyncFunc() { ... } 
const asyncFunc2 = async () => { ... }</code></pre>
                <p><strong>async</strong> functions always return a promise. It returns in one of three ways:</p>
                <ul>
                    <li>If there's nothing returned from the function, it will return a promise with a resolved value of
                        <strong>undefined</strong>.
                    </li>
                    <li>If there's a non-promise value returned from the function, it will return a promise resolved to
                        that
                        value.</li>
                    <li>If a promise is returned from the function, it will simply return that promise</li>
                </ul>
                <p>This also means that we can use normal promise syntax like <strong>.then</strong> and
                    <strong>.catch</strong> on functions that have the <strong>async</strong> keyword.
                </p>
                <pre><code class="language-javascript">const myFunc = async () => { return 5; }
myFunc.then(res => console.log(res)) // prints 5;</code></pre>
            </article>
            <article>
                <h3>The <u>await</u> Operator</h3>
                <p><strong>await</strong> is used in async functions. It is an <strong>operator</strong> that returns
                    the resolved value of a promise. <strong>await</strong> halts the execution of the async function
                    until the promise that it is used on has been fulfilled.</p>
                <pre><code class="language-javascript">const myFunc = async () => {
const res = await myPromise();
console.log(res);
}</code></pre>
                <p><strong>async...await</strong> is great because, while everything it does can be achieved with the
                    <strong>.then</strong> method, it is much cleaner and helps developers maintain and debug code.
                </p>
            </article>
            <article>
                <h3>Handling Erros</h3>
                <p>Use <strong>try...catch</strong> blocks in async functions to catch any errors/promise rejections.
                </p>
                <pre><code class="language-javascript">const myFunc = async () => {
    try {
        const res = await myPromise();
        console.log(res);
    } catch (err) {
        console.log(err)
    }
}
myFunc();</code></pre>
                <p>Alternatively, use <strong>.catch()</strong> after calling the async function. This is sometimes used
                    in the global scope to catch final errors in complex code.</p>
                <pre><code class="language-javascript">const myFunc = async () => {
    const res = await myPromise();
    console.log(res);
}
myFunc().catch(err => {
    console.log(err);
});</code></pre>
            </article>
        </section>
        <script src="./resources/prismjs/prism.js"></script>
    </main>

</body>

</html>